core/corehttp/gateway_handler.go

```diff
@@ -164,33 +164,44 @@ func (i *gatewayHandler) getOrHeadHandler(ctx context.Context, w http.ResponseWr
 		return
 	}
 
-	dr, err := i.api.Unixfs().Cat(ctx, parsedPath)
-	dir := false
+	// Resolve path to the final DAG node for the ETag
+	resolvedPath, err := i.api.ResolvePath(ctx, parsedPath)
 	switch err {
 	case nil:
-		// Cat() worked
-		defer dr.Close()
-	case coreiface.ErrIsDir:
-		dir = true
 	case coreiface.ErrOffline:
 		if !i.node.OnlineMode() {
-			webError(w, "ipfs cat "+urlPath, err, http.StatusServiceUnavailable)
+			webError(w, "ipfs resolve "+urlPath, err, http.StatusServiceUnavailable)
```

Comment:
Actually the CLI equivalent of this is `ipfs resolve -r` :) Better get it right so that the few who actually paste this into a terminal don't get confused.